File: telegram_upload_watcher/notify.py
from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass

from .queue import JsonlQueue
from .telegram import send_message
from .pools import TokenPool, UrlPool


@dataclass
class NotifyConfig:
    enabled: bool
    interval: int
    notify_on_idle: bool


def _format_elapsed(seconds: float) -> str:
    total = int(seconds)
    hours = total // 3600
    minutes = (total % 3600) // 60
    secs = total % 60
    return f\"{hours:02d}:{minutes:02d}:{secs:02d}\"


async def notify_loop(
    config: NotifyConfig,
    queue: JsonlQueue,
    url_pool: UrlPool,
    token_pool: TokenPool,
    chat_id: str,
    topic_id: int | None,
) -> None:
    if not config.enabled:
        return

    start = time.monotonic()
    last_pending = None

    await send_message(
        url_pool,
        token_pool,
        chat_id,
        f\"Watch started (elapsed {_format_elapsed(0)})\",
        topic_id=topic_id,
    )

    while True:
        await asyncio.sleep(max(1, config.interval))
        elapsed = _format_elapsed(time.monotonic() - start)
        stats = queue.stats()
        pending = stats.get(\"queued\", 0) + stats.get(\"failed\", 0)

        await send_message(
            url_pool,
            token_pool,
            chat_id,
            \"Watch status: elapsed {elapsed}, queued {queued}, sending {sending}, sent {sent}, failed {failed}\".format(
                elapsed=elapsed,
                queued=stats.get(\"queued\", 0),
                sending=stats.get(\"sending\", 0),
                sent=stats.get(\"sent\", 0),
                failed=stats.get(\"failed\", 0),
            ),
            topic_id=topic_id,
        )

        if config.notify_on_idle:
            if last_pending is None:
                last_pending = pending
            elif last_pending > 0 and pending == 0:
                await send_message(
                    url_pool,
                    token_pool,
                    chat_id,
                    f\"Watch idle (elapsed {elapsed})\",
                    topic_id=topic_id,
                )
            last_pending = pending

File: go/internal/notify/notify.go
package notify

import (
	"context"
	"fmt"
	"time"

	"github.com/nerdneilsfield/telegram-upload-watcher/go/internal/queue"
	"github.com/nerdneilsfield/telegram-upload-watcher/go/internal/telegram"
)

type Config struct {
	Enabled      bool
	Interval     time.Duration
	NotifyOnIdle bool
}

func Loop(cfg Config, q *queue.Queue, client *telegram.Client, chatID string, topicID *int) {
	if !cfg.Enabled {
		return
	}

	start := time.Now()
	retry := telegram.RetryConfig{MaxRetries: 3, Delay: 3 * time.Second}
	_ = client.SendMessage(chatID, fmt.Sprintf("Watch started (elapsed %s)", formatElapsed(0)), topicID, retry)

	lastPending := -1
	for {
		time.Sleep(cfg.Interval)
		elapsed := formatElapsed(time.Since(start))
		stats := q.Stats()
		pending := stats[queue.StatusQueued] + stats[queue.StatusFailed]
		_ = client.SendMessage(
			chatID,
			fmt.Sprintf(
				"Watch status: elapsed %s, queued %d, sending %d, sent %d, failed %d",
				elapsed,
				stats[queue.StatusQueued],
				stats[queue.StatusSending],
				stats[queue.StatusSent],
				stats[queue.StatusFailed],
			),
			topicID,
			retry,
		)

		if cfg.NotifyOnIdle {
			if lastPending >= 0 && lastPending > 0 && pending == 0 {
				_ = client.SendMessage(chatID, fmt.Sprintf("Watch idle (elapsed %s)", elapsed), topicID, retry)
			}
			lastPending = pending
		}
	}
}

func LoopWithContext(ctx context.Context, cfg Config, q *queue.Queue, client *telegram.Client, chatID string, topicID *int) {
	if !cfg.Enabled {
		return
	}

	start := time.Now()
	retry := telegram.RetryConfig{MaxRetries: 3, Delay: 3 * time.Second}
	_ = client.SendMessage(chatID, fmt.Sprintf("Watch started (elapsed %s)", formatElapsed(0)), topicID, retry)

	lastPending := -1
	for {
		if !sleepWithContext(ctx, cfg.Interval) {
			return
		}
		elapsed := formatElapsed(time.Since(start))
		stats := q.Stats()
		pending := stats[queue.StatusQueued] + stats[queue.StatusFailed]
		_ = client.SendMessage(
			chatID,
			fmt.Sprintf(
				"Watch status: elapsed %s, queued %d, sending %d, sent %d, failed %d",
				elapsed,
				stats[queue.StatusQueued],
				stats[queue.StatusSending],
				stats[queue.StatusSent],
				stats[queue.StatusFailed],
			),
			topicID,
			retry,
		)

		if cfg.NotifyOnIdle {
			if lastPending >= 0 && lastPending > 0 && pending == 0 {
				_ = client.SendMessage(chatID, fmt.Sprintf("Watch idle (elapsed %s)", elapsed), topicID, retry)
			}
			lastPending = pending
		}
	}
}

func sleepWithContext(ctx context.Context, d time.Duration) bool {
	if d <= 0 {
		return true
	}
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-timer.C:
		return true
	}
}

func formatElapsed(duration time.Duration) string {
	total := int(duration.Seconds())
	hours := total / 3600
	minutes := (total % 3600) / 60
	seconds := total % 60
	return fmt.Sprintf("%02d:%02d:%02d", hours, minutes, seconds)
}
